### PR TITLE
Feature/link vets to farmers

### PR DESCRIPTION
**Pull request by nicolasberlin**
## TL;DR
Backend to link **Veterinarian ↔ Farmer** using a **6-digit code** with a TTL.  
No UI in this PR. The UI only consumes the ViewModel and renders its state.

---

## Context
Allows connecting a vet and a farmer via an out-of-app code exchange.  
Data stored in Firestore. Handles code collisions and race conditions safely.

---

## What’s included in this PR
- **Repository**
  - `generateCode(vetId: String): Result<String>`
  - `claimCode(code: String, farmerId: String): Result<String>`
- **ViewModel**
  - `ConnectionViewModel` exposing `state: StateFlow<ConnectionUiState>`
  - Concurrency-safe: cancels any previous `Job` on `generateCode` or `claimCode`.
- **UI State**
  - `Idle | Loading | CodeGenerated(code) | Connected(vetId) | Error(msg)`
  - 
---

## How to test
1. **Read the main README** (“Connection Codes — Link Veterinarian ↔ Farmer”) for full setup details.
2. Start the Firebase emulator:
   ```bash
   firebase emulators:start
   # Firestore: 127.0.0.1:8081, Auth: 127.0.0.1:9099

---

##Important 
For local tests, the string resource has been changed:

<string name="FIREBASE_EMULATORS_URL">10.0.2.2</string>

### CODE REVIEW

**Comment by MadaMada08** [comment]
##  Summary
Thanks a lot for this solid contribution!  
The PR brings some of the backend logic I will need for implementing the Profile Screen, this is really helpfull!
I really appreciate the care taken with transactions, error handling, and test coverage.

Here are a few general remarks and questions that came to mind:

---

### What looks good to me
- The **Firestore logic** is well encapsulated, `ConnectionRepository` is clear and single-responsibility.  
- The **tests** are detailed, checking for collisions, expiration, race conditions, and edge cases.  
- Using **`StateFlow`** in the ViewModel makes the UI integration straightforward later (will be helpfull for me).  
- The **README** at the feature root is concise and helps quickly understand the purpose and flow, I think that's a really nice addition (to avoid overcrowding the code with comments also)!

---

### Points to double-check or refine
- **Firebase emulator configuration:**  
  I noticed you changed `FIREBASE_EMULATORS_URL` to `10.0.2.2`. This was the logic way of doing this week-end, since the firebase emulator was offline. However, starting 19 pm today this change will be obsolete. You should keep it only on your machine and not commit it. Please revert it to it's original state.

- **Transaction safety:**  
  In `generateCode`, Firestore transactions are retried up to 20 times on collisions I think.  
  Do you think we might want to introduce a small delay or backoff between retries to reduce contention when multiple vets generate codes simultaneously?

- **Firestore collection names:**  
  The hard-coded strings `"connect_codes"` and `"connections"` are clear, though maybe centralizing them as constants in one place (e.g. a `FirestoreCollections` object) could simplify future maintenance.

---

### Also
You forgot to run the ktfmtFormat, this is mandatory for the CI to work.
I will now add specific comments in your code where I think there are things to modify or discuss.
Also I think one or two other PRs need to be merged before yours, prepare to merge main with you code and resolve the conflicts.

Thanks again for the great work, this backend will make the connection feature easy to plug in from the UI side later on!

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    private companion object {
+        private const val CODES_COLLECTION = "connect_codes"
+        private const val STATUS_OPEN = "OPEN"
+        private const val STATUS_USED = "USED"
+    }
+
+    suspend fun generateCode(vetId: String, ttlMinutes: Long = 10): Result<String> = runCatching {
+        // Try multiple times in case of rare collisions
+        repeat(20) {
+            val code = Random.nextInt(100_000, 1_000_000).toString()
```
Comment (MadaMada08) [inline AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: Would it be worth introducing a small delay or random backoff between retries in case multiple vets generate codes simultaneously? It could help avoid bursts of transaction retries.

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        }.await()
+    }
+
+    private fun linkUsers(
+        tx: Transaction,
+        vetId: String,
+        farmerId: String
+    ) {
+        val connId = connectionId(vetId, farmerId)
+        val ref = db.collection("connections").document(connId)
```
Comment (MadaMada08) [inline AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: Would it make sense to centralize these collection names somewhere (like a FirestoreCollections object) to avoid typos or inconsistencies later?

File: AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/connection/ConnectionRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            val options = FirebaseOptions.Builder()
+                .setProjectId("test-project")
+                .setApplicationId("1:android:1")
+                .setApiKey("fake-api-key")
+                .build()
+            FirebaseApp.initializeApp(context, options)
+
+            // IMPORTANT: configure emulator on the instance before any Firestore usage
+            FirebaseFirestore.getInstance().useEmulator("10.0.2.2", 8081)
+        }
```
Comment (MadaMada08) [inline AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/connection/ConnectionRepositoryTest.kt]: You should use the same process as the other, check FirebaseEmulatorsTest. Also using the emulator was only a temporary fix.

File: AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/model/authentification/FirebaseEmulatorsTest.kt [hunk 4]
```diff
@@ -60,7 +61,7 @@ open class FirebaseEmulatorsTest(shouldInitializeEmulators: Boolean = true) {
   open fun setUp() {
     if (!emulatorInitialized && _shouldInitializeEmulators) {
       val context =
-          androidx.test.platform.app.InstrumentationRegistry.getInstrumentation().targetContext
+          InstrumentationRegistry.getInstrumentation().targetContext
```
Comment (MadaMada08) [inline AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/model/authentification/FirebaseEmulatorsTest.kt]: Could you please clarify what this change does ?

Comment (nicolasberlin) [reply AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/model/authentification/FirebaseEmulatorsTest.kt]: This change just simplifies the reference.
It replaces the fully qualified class name with a direct import of `InstrumentationRegistry`.
No behavior change — only a readability refactor.

**Review by MadaMada08 — CHANGES_REQUESTED** [review]
I see there are lots of .DS_Store files you pushed on your branch, I don't think this should end up on git.

**Comment by MadaMada08** [comment]
I will continue to re-review your code in one hour, sorry for the delay.

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                          FirestoreSchema.ConnectCodes.STATUS to STATUS_OPEN,
+                          FirestoreSchema.ConnectCodes.CREATED_AT to createdAt,
+                          FirestoreSchema.ConnectCodes.TTL_MINUTES to ttlMinutes))
+                  code
+                }
+              }
+              .await()
+
+      if (maybeCode != null) return@runCatching maybeCode
+      delay(Random.nextLong(50, 200))
```
Comment (nils-caspar) [inline AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: Why do we have to wait? and why is the delay random?

**Review by nils-caspar — CHANGES_REQUESTED** [review]
Check comments, few minor changes. Also agree with Adam that every `.DS_Store` should be removed. In the future, be careful when committing untracked files, maybe update the .gitignore to prevent such issues

File: AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/model/connection/ConnectionRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        .set(
+            mapOf(
+                "code" to code, "vetId" to user3.uid, "status" to "OPEN", "ttlMinutes" to 10L
+                // createdAt ABSENT
+                ))
+        .await()
+
+    val res = repo.claimCode(code, user1.uid)
+    assertTrue(res.isFailure)
+    assertTrue(res.exceptionOrNull()!!.message!!.contains("Missing createdAt"))
```
Comment (al23029) [inline AgriHealth-Alert-main/app/src/androidTest/java/com/android/agrihealth/data/model/connection/ConnectionRepositoryTest.kt]: Using !! can cause the test to crash if the exception is null, so it is not very safe. Prefer using assertNotNull and include a clear failure message for better debugging.

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
+package com.android.agrihealth.data.model.connection
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+class ConnectionViewModel(private val repository: ConnectionRepository = ConnectionRepository()) :
```
[the remark this answers is not in the document]
Comment (nicolasberlin) [reply AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionViewModel.kt]: I’ll note this and discuss it with the team to decide whether we should implement it.

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+          ref,
+          mapOf(
+              FirestoreSchema.Connections.VET_ID to vetId,
+              FirestoreSchema.Connections.FARMER_ID to farmerId,
+              FirestoreSchema.Connections.CREATED_AT to FieldValue.serverTimestamp(),
+              FirestoreSchema.Connections.ACTIVE to true))
+    }
+  }
+
+  private fun connectionId(a: String, b: String): String = listOf(a, b).sorted().joinToString("__")
```
Comment (al23029) [inline AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: The connectionId method uses sorted().joinToString("__") to ensure symmetry. Consider adding a comment explaining that this guarantees a consistent ID since it is not immediately obvious to readers

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+class ConnectionRepository(private val db: FirebaseFirestore = Firebase.firestore) {
+  private companion object {
+    private const val CODES_COLLECTION = FirestoreSchema.Collections.CONNECT_CODES
+    private const val CONNECTIONS_COLLECTION = FirestoreSchema.Collections.CONNECTIONS
+    private const val STATUS_OPEN = FirestoreSchema.Status.OPEN
+    private const val STATUS_USED = FirestoreSchema.Status.USED
+  }
+
+  suspend fun generateCode(vetId: String, ttlMinutes: Long = 10): Result<String> = runCatching {
+    repeat(20) {
```
[the remark this answers is not in the document]
Comment (nicolasberlin) [reply AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: I'll keep the 20 repetitions

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt (hunk 5 above)
Comment (nicolasberlin) [reply AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: The random delay prevents simultaneous retries from hitting Firestore at the same time, reducing transaction collisions when multiple codes are generated concurrently.

**Review by al23029 — CHANGES_REQUESTED** [review]
Overall:
- Both success and failure paths are clearly handled for tests, including edge cases like race conditions and missing fields.
- The use of Firebase transactions is clean and safe, especially in generateCode and claimCode.
- Naming and structure make the code easy to read and maintain.

I left four comments on areas that could benefit from small improvements. Please take a look when you have a moment. Aside from those, I ran the tests myself and everything passed without issues. Some parts were already addressed in previous reviews, and overall this was a very well-structured and easy-to-follow PR. Thank you!

File: AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+import com.google.firebase.firestore.FieldValue
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Transaction
+import com.google.firebase.firestore.firestore
+import java.time.Instant
+import kotlin.random.Random
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.tasks.await
+
+data class ConnectionCode(
```
[the remark this answers is not in the document]
Comment (nicolasberlin) [reply AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/data/model/connection/ConnectionRepository.kt]: You're right, I created for eventually use it, which is not really the case.

**Review by nils-caspar — CHANGES_REQUESTED** [review]
Improve code clarity in the test `setUp` function

**Review by nils-caspar — APPROVED** [review]
👍

**Review by al23029 — APPROVED** [review]
All the points I requested changes for in the PR have been carefully addressed. Thank you!
